trace/config.go

```diff
@@ -22,110 +22,160 @@ import (
-	ApplyTracer(*TracerConfig)
+	apply(*TracerConfig)
+}
+
+type tracerOptionFunc func(*TracerConfig)
 
-	// A private method to prevent users implementing the
-	// interface and so future additions to it will not
-	// violate compatibility.
-	private()
+func (fn tracerOptionFunc) apply(cfg *TracerConfig) {
+	fn(cfg)
 }
 
 // SpanConfig is a group of options for a Span.
 type SpanConfig struct {
-	// Attributes describe the associated qualities of a Span.
-	Attributes []attribute.KeyValue
-	// Timestamp is a time in a Span life-cycle.
-	Timestamp time.Time
-	// Links are the associations a Span has with other Spans.
-	Links []Link
-	// NewRoot identifies a Span as the root Span for a new trace. This is
-	// commonly used when an existing trace crosses trust boundaries and the
-	// remote parent span context should be ignored for security.
-	NewRoot bool
-	// SpanKind is the role a Span has in a trace.
-	SpanKind SpanKind
-}
-
-// NewSpanConfig applies all the options to a returned SpanConfig.
+	attributes []attribute.KeyValue
+	timestamp  time.Time
+	links      []Link
+	newRoot    bool
+	spanKind   SpanKind
+}
+
+// Attributes describe the associated qualities of a Span.
+func (cfg *SpanConfig) Attributes() []attribute.KeyValue {
+	return cfg.attributes
+}
+
+// Timestamp is a time in a Span life-cycle.
+func (cfg *SpanConfig) Timestamp() time.Time {
+	return cfg.timestamp
+}
+
+// Links are the associations a Span has with other Spans.
+func (cfg *SpanConfig) Links() []Link {
+	return cfg.links
+}
+
+// NewRoot identifies a Span as the root Span for a new trace. This is
+// commonly used when an existing trace crosses trust boundaries and the
+// remote parent span context should be ignored for security.
+func (cfg *SpanConfig) NewRoot() bool {
+	return cfg.newRoot
+}
+
+// SpanKind is the role a Span has in a trace.
+func (cfg *SpanConfig) SpanKind() SpanKind {
+	return cfg.spanKind
+}
+
+// NewSpanStartConfig applies all the options to a returned SpanConfig.
 // No validation is performed on the returned SpanConfig (e.g. no uniqueness
 // checking or bounding of data), it is left to the SDK to perform this
 // action.
-func NewSpanConfig(options ...SpanOption) *SpanConfig {
+func NewSpanStartConfig(options ...SpanOption) *SpanConfig {
 	c := new(SpanConfig)
 	for _, option := range options {
-		option.ApplySpan(c)
+		option.applySpanStart(c)
 	}
 	return c
 }
 
-// SpanOption applies an option to a SpanConfig.
+// NewSpanEndConfig applies all the options to a returned SpanConfig.
+// No validation is performed on the returned SpanConfig (e.g. no uniqueness
+// checking or bounding of data), it is left to the SDK to perform this
+// action.
+func NewSpanEndConfig(options ...SpanEndOption) *SpanConfig {
+	c := new(SpanConfig)
+	for _, option := range options {
+		option.applySpanEnd(c)
+	}
+	return c
+}
+
+// SpanOptions applies an option to a SpanConfig. These options are applyable
+// Only when the span is created
 type SpanOption interface {
```

Comment:
```suggestion
type SpanStartOption interface {
```

Reply:
I didn't take this change for two reasons.
1. This creates a lot of churn in our tests.
2. These options are about modifying a `spanConfig`, and every field in the config is represented by a SpanOption.  
 
If anything I think it would be more acceptable to move SpanEndOptions to a unique config, `spanEndConfig` maybe.

Reply:
The reservation I have here is that we need to include information in the description about what this is used for instead of having the name indicate its use. From the end user trying to stand up an SDK the one-to-one mapping to the SpanConfig isn't going to be known and I'm not sure we need to stress the mapping between the config name an option name.

I'm not sure it's a huge deal, but there will be an asymmetry not having a "start" option to compliment the "end " option. This as well might lead to some confusion on the end users' journey, but I imagine their IDE can help with this.

If we think the change would be too intrusive here we can rename it in a follow on PR.